mlflow/store/tracking/sqlalchemy_store.py

```diff
@@ -589,28 +591,76 @@ def log_metric(self, run_id, metric):
             self._check_run_is_active(run)
-            # ToDo: Consider prior checks for null, type, metric name validations, ... etc.
-            logged_metric, just_created = self._get_or_create(
-                model=SqlMetric,
-                run_uuid=run_id,
-                key=metric.key,
-                value=value,
-                timestamp=metric.timestamp,
-                step=metric.step,
-                session=session,
-                is_nan=is_nan,
-            )
-            # Conditionally update the ``latest_metrics`` table if the logged metric  was not
-            # already present in the ``metrics`` table. If the logged metric was already present,
-            # we assume that the ``latest_metrics`` table already accounts for its presence
-            if just_created:
-                self._update_latest_metric_if_necessary(logged_metric, session)
 
-    @staticmethod
-    def _update_latest_metric_if_necessary(logged_metric, session):
+            # commit the session to make sure that we catch any IntegrityError
+            # and try to handle them.
+            try:
+                self._save_to_db(session=session, objs=metric_instances)
+                self._update_latest_metrics_if_necessary(metric_instances, session)
+                session.commit()
+            except sqlalchemy.exc.IntegrityError:
+                # Primary key can be violated if it is tried to log a metric with same value,
+                # timestamp, step, and key within the same run.
+                # Roll back the current session to make it usable for further transactions. In the
+                # event of an error during "commit", a rollback is required in order to continue
+                # using the session. In this case, we re-use the session to query SqlMetric
+                session.rollback()
+                # obtain the metric history corresponding to the given metrics
+                metric_history = (
+                    session.query(SqlMetric)
+                    .filter(
+                        SqlMetric.run_uuid == run_id,
+                        SqlMetric.key.in_([m.key for m in metric_instances]),
+                    )
+                    .all()
+                )
+                # convert to a set of Metric instance to take advantage of its hashable
+                # and then obtain the metrics that were not logged earlier within this run_id
+                metric_history = {m.to_mlflow_entity() for m in metric_history}
+                non_existing_metrics = [
+                    m for m in metric_instances if m.to_mlflow_entity() not in metric_history
+                ]
+                # if there exist metrics that were tried to be logged & rolled back even though
+                # they were not violating the PK, log them.
+                if non_existing_metrics:
+                    self._log_metrics(run_id, non_existing_metrics)
```

Comment:
On second thought, reading up to 1000 metric histories into memory for conflict resolution could be fairly expensive from a latency / memory pressure perspective.

I'm wondering if it might be better to throw in this case instead. For certain sql dialects, we can use a "on duplicate key ignore" construct and otherwise fail open. I'll update this with some more details shortly.

Thoughts @erensahin ?

Reply:
The more I think about it, the more I favor the current approach. The data volume concern can be addressed via further chunking

Reply:
Thank you so much for handling the problem @dbczumar :)

In my opinion, throwing an error when duplicate metric entry (key-timestamp-step-value) is given in a batch is perfectly fine without even trying to insert. It is the way how log_params treat. And I agree on "on duplicate key ignore" approach on certain dialects, but for the sake of stability raising an early error makes more sense for me.

Reply:
Understood! I was tempted to go that direction, but I don't want to break any existing user workloads that might rely on idempotent logging of duplicate metric keys. I've just made a minor adjustment here to chunk the metrics into batches of 100 to avoid loading too much data into memory at once; I think the data volume is acceptable since this case is unlikely to occur frequently. I also refactored slightly so that we call an internal method when handling the IntegrityError, rather than retrying the whole `_log_metric` method, which would create a nested session. I've gone ahead and pushed the changes; let me know how they look! If CI passes, I think we're all set to merge! Thanks for all your hard work @erensahin !